### ISSUE TITLE
simple editor for text styles

### ISSUE THREAD

**Issue by lexoyo** [silexlabs/Silex#52]
Being able to edit the styles available in the text editor, i.e. Title, Subtitle, Title 1, Title2, ... Normal text, Quotes

And also add other styles

**Comment by budislav** [comment]
I can't wait to see this!
But I can't imagine how this will be written in css and html doc without some things I saw in grapesjs for example.

![untitled](https://user-images.githubusercontent.com/3619927/36710972-4f3834fe-1b81-11e8-9281-cf238d195b8c.png)

Can you develop this way of adding classes and naming them, so other components of style editor can work on the same way? On this way we can know where to find this styles in css or what is the name of div in html. I think this is really necessary. And this can be used for all other styles in future.

**Comment by lexoyo** [comment]
Hello

As I told you before, I'm working on this, and as you asked, here is what I am up to

![screenshot from 2018-03-07 19-57-43](https://user-images.githubusercontent.com/715377/37112136-e90120a2-2241-11e8-8798-30b1c226fb16.png)
![screenshot from 2018-03-07 19-51-28](https://user-images.githubusercontent.com/715377/37111913-4c9df154-2241-11e8-997b-1a7e022ebb66.png)
![screenshot from 2018-03-07 19-50-25](https://user-images.githubusercontent.com/715377/37111914-4cba2cde-2241-11e8-920a-a1fe2d71bbdd.png)

Concerning the CSS classes, here is what I managed to do without taking too long:

![screenshot from 2018-03-07 19-59-13](https://user-images.githubusercontent.com/715377/37112190-0c4bd548-2242-11e8-97bd-7ae1115031ac.png)
![screenshot from 2018-03-07 19-59-01](https://user-images.githubusercontent.com/715377/37112192-0c6d8dfa-2242-11e8-9ebe-281291852a7e.png)

Feel free to ask questions
The buttons at the bottom are for reset/copy/paste the style but I'm not sure if it is useful yet
The wole system is a prototype, I'm not sure how useful it is

**Comment by budislav** [comment]
I am happy to see this nice progress!

I believe the reason why groups of text settings (title 1,2,3, links, paragraph) are all visible under the same tab is because it is still a prototype?
I think the best is this settings to be shown up under style tab as one accordion button, also only one group of settings are needed. So when we select title or paragraph or links on the canvas, one group of settings should be shown up depending on what we selected but not all time visible every group of same settings. I believe that links, titles and paragraph have same settings?

Reset/copy/paste are really needed I just think that is not right place for them.
So in case we have text options under style tab, buttons (Reset/copy/paste) could be placed on text accordion button on the right side on hover.

Thank you for working on this really important part! Keep updating us :)

**Comment by lexoyo** [comment]
Thank you for your questions and suggestions @budislav 

Before I can answer, here is where I am now:
* I have found a way to do the "adding classes and naming them" thing, but the Silex way, less visually appealing than grape's but it works for people who don't understand css
* the style editor is no longer linked to the selection on stage, it is about global styles and CSS classes
* I found a way to add a "state" which makes it possible to style texts for desktop or mobile differently

Here is a demo of the style editor in its current state:

Change color of all texts: all website, all text boxes, all styles normal, Title 1, Title 2, Title 3
![style-editor-all-texts](https://user-images.githubusercontent.com/715377/37243722-a5cdfeec-247e-11e8-82cb-051b2723efef.gif)

Change color of all "Title 1" texts

![style-editor-h1](https://user-images.githubusercontent.com/715377/37243727-acd6866e-247e-11e8-891f-b292c7dddce7.gif)

Change color of the texts in 1 text box

![style-editor-css-class-speedup](https://user-images.githubusercontent.com/715377/37243760-0cdc4a9e-247f-11e8-98be-c708ee551696.gif)

Reset a style

![style-editor-reset](https://user-images.githubusercontent.com/715377/37243732-b4a58b06-247e-11e8-8f42-f1f18d46b964.gif)

Change color of all the texts in 1 specific section

![style-editor-css-section](https://user-images.githubusercontent.com/715377/37243772-2affc6fe-247f-11e8-8f80-644cfadcd6db.gif)

Make text bigger on mobile for the whole website

![style-editor-mobile-speedup](https://user-images.githubusercontent.com/715377/37243859-a682c7f8-2480-11e8-85ea-219f46dcdb14.gif)

**Comment by budislav** [comment]
Thanks for demos @lexoyo very nice!
Looks very promising, keep going!

**Comment by lexoyo** [comment]
Hello
I need feedback about where I am now. Not everything is working, only a part of it.

Here is the tests you can do
* open https://silex-test-tmp.herokuapp.com/
* select a blanck template
* add a text box to a section somewhere
* in the text box, put some text, use only "Normal", "Heading 1", "Heading 2", "Heading 3" and add links too, **NO color or bold inside the text editor please**
* click the "style editor tab" in the property panel
* use the editor to style texts for desktop or mobile, hover state, other states

Please ask questions about how the style editor works, this will let me understand how you see things

cc @budislav @ceubri @BaguetteAgile

**Comment by ceubri** [comment]
Application error...

**Comment by lexoyo** [comment]
oops.. fixed :)

**Comment by lexoyo** [comment]
Nice one Ceubri!
You can do exactly that with the style editor: select the body (ESC key) then change H1 style
Or do you mean we should be able to style all H1-H3 but not P at once?

**Comment by ceubri** [comment]
Ok it's working with the body stylization, nice and powerful !

> edit "all headings" could be usefull to change the font familly or to capitalize all headings (not p) at the same time, but it's not easy to integrate this in the UI... or maybe add "titres" in the editor ?

My impression : i'm quickly lost, i don't remember what's styled or not.
Maybe it could be easier not to have to select a container or anything, just have the basic (paragraph, Titles 1,2, 3...) and have the possibility to add new ones, for example when the section is colored (section-Title1, section-Title2,...), 

I thing it's normal because you'r workin on:
> duplicate, add, remove, reset are not working
> how to remove a background-color
> how to apply a style

**Comment by lexoyo** [comment]
Very interesting thx
Did you notice the "*" in the state and visibility names? It means that there is a style there... We could do the same with H1-H3 names and even the properties
Does it help?
Maybe you are right, selection may be too complex? But it means that you can not simply style a text box without understanding styles...

**Comment by lexoyo** [comment]
So @ceubri what you say is that the style editor's ability to define styles for each element is confusing because you will not remember where you have put what style, right? I have had the same feeling and this is why every time you define a style on an element is is "remembered" and listed in the style drop-down list. Also the visibility list and the states list have a "*" in the names when a style has values defined for a visibility or state. Is that clear for you @ceubri and others? Does that help? How to make it easier to understand?

Ceubri you suggest it would be easier to have only named styles: on for the body and others that you define and apply to some elements. This is a very good simplification since selection does not change the editor anymore. The problem with that is that if a user does not understand the style system and just want to change the color of the text in the selected text box, it is only possible thanks to the buttons in the text editor. My idea with the style editor was to remove the buttons of the text editor and leave only H1-H3, P, links, and something like B or strong. I like this idea because there is only one style editor to maintain... But maybe it is good to keep a simple style editor in the text editor (only font, size, color) and the style editor for advanced styling. Then your solution seems better @ceubri... Another solution is to do as I did (style editor only, selection creates styles) but to limit styles to body, sections and text boxes. So 3 solutions
* As is: selection creates styles
* As is but with styles only for body, sections, text boxes (not containers)
* Text editor with simple styling and style manager only for named styles

I will think about it but your solution Ceubri (number 2) may be the best one...

**Comment by lexoyo** [comment]
There is something else I did not anticipate: in the curent style editor, if I first define the style of a text box, then the style of it's container, the style which will be applied is the one defined on the container whereas I would prefer the text box one. This is because the last one I define is the last in the Dom... So whatever solution we choose, we may need to display the styles in a reorderable list instead of a drop-down list. The first styles would be "stronger" than the last ones. This is hard to understand for the user right?
Another solution would be to arbitrary decide that body is weaker, then sections, then containers, then text boxes...

**Comment by lexoyo** [comment]
Or the deeper the stronger would be good

But for named styles, these 2 last solutions can not be applied since a named style can be applied to a container, a section and a text box simultaneously :-/

**Comment by ceubri** [comment]
I think : 
- it is good to keep a simple style editor in the text editor (only size, color, positionning... I'm not sure font family is good there) and the style editor for advanced styling (save styles)

- or if you keep only 1 editor, you'll have to disable the "double click" fonction to open the basic editor.
> click on text = selection
> click on text border = change crop
> click on corner = displace
and then you have only 1 editor (top the basics, bottom the advanced style editor), and you style only texts

- is it normal not to have a "save button" in the editor ?

**Comment by lexoyo** [comment]
> is it normal not to have a "save button" in the editor ?

As it looks like a panel I would say yes it is... Each change is applied immediately. And it is undoable

Would you expect a save button?

> it is good to keep a simple style editor in the text editor (only size, color, positionning... I'm not sure font family is good there) and the style editor for advanced styling (save styles)

This is feasible, based on the current text editor. The main issue is that wysiwyg text editors work poorly and have bad markups - maybe it's me but I never managed to have a good one. Of we only display styles there, it is easier if think. The user enters text with no colors or any formatting and silex makes the styling

But still it is possible

> or if you keep only 1 editor, you'll have to disable the "double click" fonction to open the basic editor.

That's a good idea
You mean edit text in place not in an editor right?

**Comment by budislav** [comment]
This is how I think it should work and it is the easiest way also other web editors like webflow works:
* Text should not be editable trough wysiwyg text editor, but directly from style editor tab. So no double click on text box to see whole text. We need those settings in style editor tab.
* Designing a modern web site consider placing titles in other div's or places title and text are not linked together always.  There should be insert Title and insert text in add element panel.
* There should be also add link in add element panel.
* Making a style is really great idea you can make style by editing properties, copy that style and apply on other same element. Please don't remove this option. But we need possibility to change that style name (class) from class name.
* In style editor properties should have only settings for element you select on canvas not to have always visible settings for paragraph, link, title, title1, title2, title3. but if you select title element you have added trough add element panel on canvas you will see only setting for paragraph + buttons from wysiwyg text editor I mentioned earlier. 

I will make a mockup so you understand everything easily.

**Comment by lexoyo** [comment]
>   * Text should not be editable trough wysiwyg text editor, but directly from style editor tab. So no double click on text box to see whole text. We need those settings in style editor tab.

Ok, I will do this soon, in a 2nd step

>   * Making a style is really great idea you can make style by editing properties, copy that style and apply on other same element type. Please don't remove this option. But we need possibility to change that style name (class) from class name. Style name is actually class name, css should be clean and understandable. So if you like something on canvas you make for example you see title somewhere on canvas you like it, you just select it from canvas click on copy style from style editor, then select new title element you like to apply on that style on canvas and click on paste style. It can not be simpler.

Ok I will do this now, 1st step

>   * Designing a modern web site consider placing titles in other div's or places title and text are not linked together always. There should be insert Title and insert text in add element panel. So two separated elements.
>   * In style editor properties should have only settings for element you select on canvas not to have always visible settings for paragraph, link, title, title1, title2, title3. but if you select title element you have added trough add element panel on canvas you will see only setting for paragraph + buttons from wysiwyg text editor I mentioned earlier.

Concerning these 2 points, I see exactly what you mean. The problem I have is linked to Silex's choice to have everything absolute by default. So create one H1 block then one H2 block and then a normal text block is not possible. One would have to align them by hand, and when one block is changed and has more or less lines, one would have to move all the blocks bellow it up or down...

So 1 single block for titles and text is needed in Silex. 

This is also good in case you have dynamic text coming from translation system (coming soon) or a CMS

**Comment by budislav** [comment]
> Concerning these 2 points, I see exactly what you mean. The problem I have is linked to Silex's choice to have everything absolute by default. So create one H1 block then one H2 block and then a normal text block is not possible. One would have to align them by hand, and when one block is changed and has more or less lines, one would have to move all the blocks bellow it up or down...

But nobody code site with absolute values today. I never seen a whole website  made with absolute sizes. It is hard do to make responsive websites since you need to touch every element and move it for every media query. This absolute positioning is problem for fluid webdesign.

Can we consider making css grid system? CSS grid is presented in every browser today, there are no reason why not to use them. Also since nobody make this in their editor, Silex can be first to have something unique and it can gain popularity only with this. Also responsive designing should be automatically done then.

**Comment by BaguetteAgile** [comment]
>  Concerning these 2 points, I see exactly what you mean. The problem I have is linked to Silex's choice to have everything absolute by default. So create one H1 block then one H2 block and then a normal text block is not possible. One would have to align them by hand, and when one block is changed and has more or less lines, one would have to move all the blocks bellow it up or down...


Bonjour, je viens de jeter un oeil au projet, super beau travail. Seulement il est vrai que j'ai difficile à retrouver les Styles de titre que j'ai mis pour ensuite pouvoir les éditer. j'ai pas bien compris le post avec les "*" 
j'essai de tester plus en profondeur en fin de semaine, je continu de suivre l'évolution.


**Hello, I just took a look at the project, great work. Only it is true that I have difficult to find the title Styles that I put to then be able to edit them. I did not understand the post with the "*"
I tried to test more in depth at the end of the week, I continue to follow the evolution.**

Good Work

**Comment by lexoyo** [comment]
> But nobody code site with absolute values today. I never seen a whole website made with absolute sizes. It is hard to make responsive websites since you need to touch every element and move it for every media query. This absolute positioning is problem for fluid webdesign.

This is what is done in wix and silex historically. I believe it is much easy for real beginners to create websites. This vision has been challenged several time when I re-started silex from scratch, and it will eventually come back unless I have time to implement this grid system which you are pushing for some time now @budislav and which is clearly a great idea. It is also a good amount of work but I definitely keep it in mind.

> Hello, I just took a look at the project, great work. Only it is true that I have difficult to find the title Styles that > I put to then be able to edit them. I did not understand the post with the "*"
> I tried to test more in depth at the end of the week, I continue to follow the evolution.

Thank you baguette! This is useful feedback. The "*" means that values have been set. It is meant to help you remember where you have changed things...

**Comment by lexoyo** [comment]
Here is what I did this afternoon and now live on https://silex-test-tmp.herokuapp.com/

* [x] selection on the stage does not change what style is being edited
* [x] you can create styles and apply them to multiple text boxes
* [x] if you use the base template I did, you will have a "all texts" style to style all texts at once
    => to test with my template, go [here](https://github.com/lexoyo/style-editor-test) and click "fork", then go back to [silex here](https://silex-test-tmp.herokuapp.com/) and open github / style-editor-test / master / empty.html and then add text boxes, play with styles

What I did not do yet
* UI / design the tool
* I did not change the text editor yet (step 2 of budislav and ceubri 's feedback)

Let's keep thinking and talking about that, I feel like we are getting closer

**Comment by budislav** [comment]
@lexoyo I think it should apply style automatically on element when you select style from select button so we don't need apply current style button. Because we already named our style we can select it back from list. Easier and faster. Silex should automatically set name for style but user can change it.  When you select text box from canvas style from select buton should change to that style that is applied to that text box. So easily you can see what style is when you select each text box from canvas. Also properties will change with style.

**Comment by BaguetteAgile** [comment]
Very good job, is it possible to add (see the screenshot) information that gives us directly the style used for the current selection, such as:

My style
H1

as you can easily navigate the menu to change the current selection.

Otherwise I did not notice any apparent bug ...

I'm still watching this week

![capture du 2018-04-04 20-30-57](https://user-images.githubusercontent.com/35597855/38327664-1418bf30-3849-11e8-9ec2-f8912020657c.png)

**Comment by budislav** [comment]
So this is how I think it should look like.
Visibility select button should be under settings tab.
I removed blink from decoration because it is not supported everywhere?

![text_style_editor](https://user-images.githubusercontent.com/3619927/38328924-7b23c8e8-384c-11e8-9b98-0f78834ce339.png)

**Comment by lexoyo** [comment]
Thank you @all for your participation (welcome @jaellozano @valentinogagliardi @AndreaMonzini)

> * It should apply style automatically on element when you select style from select button so we don't need apply current style button. Because we already named our style we can select it back from list. Easier and faster.
> * Silex should automatically set name for style but user can change it. 
> * When you select text box on canvas, select style buton with their settings should change to that style that is applied to that text box. So easily you can see what style is applied when you select each text box from canvas.

It looks like what I did before right? It was a bit confusing right? Could you describe each thing I have to change to do that? 

> Very good job, is it possible to add (see the screenshot) information that gives us directly the style used for the current selection, such as: My style, H1

Would it be ok to have a "*" where values are set beside the H1-H3 names? Or do you mean a field which says "selection style: My Style" maybe? Or maybe is it as @budislav suggests, to select the style of the element which you select on the stage?


> @budislav: this is how I think it should look like.

I love it when you do that!!!
I see you added list style, which is a great idea
You removed the selection of <H1>, <H2>, <H3>, <P>, <A>? As discussed before, I am aware it is not the ideal solution but we need that for the moment... Maybe beside "pseudo class" as a small drop down list?

Also I wanted to add <b>, <i> and <u> to be able to style the elements marked by the user in the text editor. And a "all" to edit styles of all the titles and texts for the selected style.  

> Visibility select button should be under settings tab.

Ok, good point! It could be the same as for the properties: when you are in Desktop view then it changes the styles for desktop+mobile and in mobile view it changes styles for mobile only, what do you think? Will it be clear for the user? Should we display it somewhere?

> I removed blink from decoration because it is not supported everywhere?

:+1: 

> You can change to numerical font size by clicking on that w letter that stays for word. I don't know how to name that mode with words in sizes. So you click on that w and it should dorp list with px, em and %.
This is also one step closer to have other css style properties so on the end we have one tab for all styles, just like in Webflow.

* [ ] Ok, this I will do in a 3rd step ok? I will think about how to do that in prodotype

In the mean time I will keep the 2 values (word/numeric) and if one is defined the other has no effect

All feedback are welcome!

**Comment by lexoyo** [comment]
Oh and the pseudo class could be display like in the browsers developer consoles, with a dropdown marked as a ":hov" button
![dev-tool-hover1](https://user-images.githubusercontent.com/715377/38362934-ea9cbc1c-38d2-11e8-8572-b278a6fb56ec.png)

Or like this maybe

![dev-tool-hover2](https://user-images.githubusercontent.com/715377/38362941-f0931580-38d2-11e8-860d-3b1699769676.png)

**Comment by ceubri** [comment]
I just finish my test and...
... Great job Lexa, i like your style editor !

**Comment by budislav** [comment]
> It looks like what I did before right? It was a bit confusing right? Could you describe each thing I have to change to do that?

I don't quite remember how is working before you changed this fastly :D 

> * It should apply style automatically on element when you select style from select button so we don't need apply current style button. Because we already named our style we can select it back from list. Easier and faster.

The process is next:
1. You put a text box on canvas
2. Style (class) is automatically added named "textbox1" and applied so you can edit text styles immediately. You can rename it by pressing the button I made a pencil symbol but renaming is optional. Next text boy will be automatically get name textbox2 and so on.
3. Double click on it to edit text and it don't open text editor but you can just type inside of that text box. We have all text style settings in the style panel. You can change style settings and style need to be automatically applied without pressing any button. You already elected that text box so I know what I want to edit right away.
4. Every change you made in style editor is saved to that style for example textbox1 or how you renamed it. So if you want to change style for that text box to some style you made before you just select style from style select button and style should be automatically applied. On that way you can easily switch between styles and see directly on canvas all changes. We don't need apply button.

>* Silex should automatically set name for style but user can change it.

I explained that above.

>* When you select text box on canvas, select style button with their settings should change to that style that is applied to that text box. So easily you can see what style is applied when you select each text box from canvas.

Also explained this above. When you select text box on canvas it will select style in select button and show all style settings for style we made for that textbox. You will see style settings of every text box you select on canvas in style panel.

> I love it when you do that!!!

I want we make serious application for everyday working and making serious stuff, no focus on amateurs anymore please :)

> Would it be ok to have a "*" where values are set beside the H1-H3 names? Or do you mean a field which says "selection style: My Style" maybe? Or maybe is it as @budislav suggests, to select the style of the element which you select on the stage?

Just keep in mind simple and effectively. I need to work more on this problem.

> You removed the selection of `<H1>,` `<H2>,` `<H3>,` `<P>,` `<A>` ? As discussed before, I am aware it is not the ideal solution but we need that for the moment... Maybe beside "pseudo class" as a small drop down list? 

I will add those things in next mockup I am working right now. This also need more thinking to make it as simple as possible.

> Also I wanted to add `<b>,` `<i>` and `<u>` to be able to style the elements marked by the user in the text editor. And a "all" to edit styles of all the titles and texts for the selected style.

This should be done smartly, I need to figure out, but please no more text editor. That's why we have now everything for text in style panel. That text editor should die. Kill him please.

> Ok, good point! It could be the same as for the properties: when you are in Desktop view then it changes the styles for desktop+mobile and in mobile view it changes styles for mobile only, what do you think? Will it be clear for the user? Should we display it somewhere?

Definitely! I need to work on this.

> You can change to numerical font size by clicking on that w letter that stays for word. I don't know how to name that mode with words in sizes. So you click on that w and it should dorp list with px, em and %.
Ok, this I will do in a 3rd step ok? I will think about how to do that in prodotype
In the mean time I will keep the 2 values (word/numeric) and if one is defined the other has no effect.

This only applies to that input for changing font size, because there is only one type you can have at same time, so it should be one input and you can just change it by selecting it from that letter in the corner.

I think you should wait to see whole concept before you change anything :)

**Comment by budislav** [comment]
@lexoyo 
> Oh and the pseudo class could be display like in the browsers developer consoles, with a dropdown marked as a ":hov" button
dev-tool-hover1

So this means there can be more activated more than one?
I like how Webflow made that, what do you think?
![wf states](https://user-images.githubusercontent.com/3619927/38373604-eec13f1a-38f1-11e8-9d66-23cde2ad1db6.png)

**Comment by lexoyo** [comment]
> So this means there can be more activated more than one?

Only 1 at a time

>  like how Webflow made that, what do you think?

Yes it looks nice!!
Should it be reset when we change the style to edit? How will the user know which state is being edited?

**Comment by budislav** [comment]
> Should it be reset when we change the style to edit?

Yes I will add a button for reset.

> How will the user know which state is being edited

We can color the name of those settings are changed.

**Comment by budislav** [comment]
Ok, I think I touch everything.
Since we have possibility to save text css properties to style, we can save other things too. I know we don't have all of them right now but we need to have a space for them and this style tab is excellent.
We need also number of how much element with same style are on the same page or others. We can select that circle or text and they will be selected on canvas.

![styles](https://user-images.githubusercontent.com/3619927/38387597-a236010e-3918-11e8-941f-c7632886ed45.png)

On the top there is a name of html tag we edit and in this case it is h3 which we can select under component settings tab.

![settings](https://user-images.githubusercontent.com/3619927/38387662-d7ced6ec-3918-11e8-9c4f-6234c9bcdc92.png)

They are still text box but you can set exactly what you want it to be. On this way we will have like separated elements. They are still absolute positioned because of Silex background nature right now.

For applying bold italic or link on word on canvas the best it would be if it looks like this.

![edittext](https://user-images.githubusercontent.com/3619927/38387840-8f00ee40-3919-11e8-9635-d1562a01e6f3.png)

So when you set a link for example it should look like this

![link](https://user-images.githubusercontent.com/3619927/38387857-9e954720-3919-11e8-8559-c5b025c6531a.png)

**Comment by lexoyo** [comment]
Oh wow this is awesome!

It will take several steps to get there but let's do something like this

Starting with the UI.... Could you create png images with transparent background and white icon for style, variant, decoration, transform please? The others I might have them in font awesome?

**Comment by budislav** [comment]
I can make .png, do you maybe want .svg so you put those into font icons for HDPI screens?
Icons should be scalable so they look nice on retina screens.

One icon I made wrong. Under style there are oblique option, which is similar to italic, I don't see a difference, and I don't know how to design that icon.
https://www.w3schools.com/cssref/tryit.asp?filename=trycss_font-style
https://stackoverflow.com/questions/1680624/font-style-italic-vs-oblique-in-css

Bold icon I place there by mistake should be on canvas inline editing under selected text as I show you on images above, because bold is used when you want to bold only one word or more, but for whole sentence usually works changing font weight.

**Comment by lexoyo** [comment]
Sure, svg is even better

Oblique we can just forget right? Then style becomes an on/off button? Or a check box?

Concerning bold, it will be a button in the text editor tool bar, whereas font weight is supposed to be in the style editor. Should I do it with a button bar or a drop-down list?

**Comment by budislav** [comment]
@lexoyo 
Yes we can forget about oblique.

Style is just without oblique button like in mockup.

Font weight would be ideal to be in shape of drop-down button like in mockup. (I have updated image in that post above).

Also it would be awesome if everything you hover in drop-down list shows on canvas directly like preview, so you don't need to click on each option to see how it will look, but it will activate it only on click of course.

**Comment by lexoyo** [comment]
Hello

A quick note about where I am today on this subject

I have added several things in Prodotype, the project which holds the UI for Silex. It looks like this
![screenshot from 2018-04-23 09-43-18](https://user-images.githubusercontent.com/715377/39113166-07228310-46db-11e8-964b-d3d9636b06fb.png)

The push buttons with icons are there to create the UIs you specified @budislav 
Please [try these UI components on Prodotype demo page](http://projects.silexlabs.org/Prodotype/pub/)

**Comment by budislav** [comment]
Glad to know you are working on UI! @lexoyo

**Comment by lexoyo** [comment]
This is my "road map" for the weekend (very ambitious but I have a lot of coffee)
* [ ] the tags which we can "style": H1, H2, H3, P, B, I, U, A
* [ ] integrate @budislav 's icons and UI
Add link-active class and other [Silex specific css classes](https://github.com/silexlabs/Silex/wiki/Silex-CSS-editor)
* [ ] fix buttons bugs (rename, create, ...)

**Comment by BaguetteAgile** [comment]
Good luck with the work  !!

**Comment by budislav** [comment]
let be easy to implement! :)

**Comment by lexoyo** [comment]
Hi guys, thx for the support :)
Here is what I did in approx 10 hours, please test with the "responsive" blank template only
https://silex-test-tmp.herokuapp.com/

![style-editor-step1](https://user-images.githubusercontent.com/715377/39399319-f0d540ee-4b1b-11e8-8b0c-4ee1b55e3571.gif)


Notes and questions:
* the buttons under the style name are not design and have bugs
* the "mobile" icon on the left of the style name means: "mobile only" and it should probably be linked to the mobile mode (enter/exit the mobile editor when one press this button)

As always, I'll appreciate your tests, questions and feedback :)
You're the bests

**Comment by lexoyo** [comment]
I'm thinking also I should
* [ ] in font-size: only keep the numeric version, not the named sizes (xx-small...)
* [ ] add the list styles
* [ ] add margins and paddings which are useful not only for text... so maybe it should be put somewhere else?

Did you see how powerful the psheudo classes are?
![screenshot from 2018-04-28 19-53-47](https://user-images.githubusercontent.com/715377/39399442-e7279aa4-4b1d-11e8-98ae-90ad5b4b4cb4.png)

**Comment by lexoyo** [comment]
UI integration is really exhausting

But it starts to look good:
![screenshot from 2018-04-29 19-06-22](https://user-images.githubusercontent.com/715377/39409065-a4dc0f9a-4be0-11e8-923d-7d30e99c7cfb.png)

I used what I have done in Prodotype + the icons and UI ideas of @budislav 
And also
* [x] in font-size: only keep the numeric version, not the named sizes (xx-small...)
* [x] add the list styles
* [x] Add link-active class and other Silex specific css classes

Now live on https://silex-test-tmp.herokuapp.com/

**Comment by budislav** [comment]
Hi @lexoyo 
Thank you for your hard work here!
I need to explain my mockup little more. There should be only Paragraph, heading, Link and List and what you select you will see new options under them, just like in mockup I made. Also this should be under component settings tab, text box is one component, it's a modular system.

![heading](https://user-images.githubusercontent.com/3619927/39410275-2da8399c-4bf6-11e8-8f10-89e5d1131012.png)

When headings are selected there are options: h1,h2,h3,h4,h5,h6.
When List is selected there are options: ul, ol.
Paragraph don't have any options and for link I should design that mockup.

Bold, underline and Italic are styles for text those buttons should be under style tab. Also on canvas.

Also that mobile icon should be only indicator that will show on what mode you are, so when you are making a style for mobile you need to select mobile editor button to go into that mode, then everything you change from style will be on visible mobile only. I don't understand this concept you made now with mobile only?

> In font-size: only keep the numeric version, not the named sizes (xx-small...)

Agree.

> add the list styles

We can select style in that dropdown select button. I would say don't focus on that right now.

> add margins and paddings which are useful not only for text... so maybe it should be put somewhere else?

Yes we need, that's why I made that accordion button "Layout" above Typography :) Just look at Webflow Layout section, that is just what we need.

![screenshot from 2018-04-29 21-59-04](https://user-images.githubusercontent.com/3619927/39410494-79fb769e-4bf9-11e8-9b14-2e09cdd8b572.png)

**Comment by lexoyo** [comment]
I'll answer you very soon @budislav 
I have almost finished the integration
https://silex-test-tmp.herokuapp.com/
![screenshot from 2018-04-30 18-01-17](https://user-images.githubusercontent.com/715377/39437216-834fc50e-4ca0-11e8-9f22-6b6b3812ecc7.png)

**Comment by lexoyo** [comment]
I think there is something wrong in the style editor: if one want to change all the texts from black to white, one has to select P and change the color to white, then do the same with H1, H2, H3, A, ...

The solutions I see:
1- keep it like that, it is not an issue
2- add a "All" or "*" to change all the tags at once
3- change P to "All or *"
4- make it possible to select multiple tags (P and H1, H2, H3) and then when one changes the color, all of them are changed

What do you think?

**Comment by BaguetteAgile** [comment]
> 4- make it possible to select multiple tags (P and H1, H2, H3) and then change the color, all of them are changed

Hello

I think it would be more interesting to have the possibility to choose several tag (and not necessarily "all") to change the colors of the text.

When we create a text for its website we do several color tests to find the best and as such we have the opportunity to make a live comparison with the different paragraphs or titles.

I find solution 4 the best.

the multiple selection is possible with such a key combination as
Ctrl + click?

**Comment by budislav** [comment]
I must admit I don't understand how this styling working. That's why I thought it should be the simple as possible and to use presented system we already can see how it works  (Webflow for example) so it can be a model for Silex.
I think it should automatically apply new style when add new text box, also when select text box on canvas it should show selected style for that box in style editor.

**Comment by lexoyo** [comment]
> I need to explain my mockup little more. There should be only Paragraph, heading, Link and List and what you select you will see new options under them, just like in mockup I made. 
> When headings are selected there are options: h1,h2,h3,h4,h5,h6.
When List is selected there are options: ul, ol.
Paragraph don't have any options and for link I should design that mockup.

That is a good way of seeing things! I missed that point... It seems complicated to me right now, let me think about it

Also, do you think I should add H4 to H6 ? Or is H1 to H3 enough ?

> Also this should be under component settings tab, text box is one component, it's a modular system.

Ok, I start to see that now. It will come at one point but for now the 2 tabs (text styles and properties) are made with a different technology (my other project "Prodotype" for the text styles, and pure JS in silex for the properties). So be patient, we'll get there :)

> Bold, underline and Italic are styles for text those buttons should be under style tab. Also on canvas.

Ok this is interesting too... I see them as a way to style part of texts: unlike H1-H3, P and lists, the U, B and I tags can be applied to a portion of the text, not a whole line. So in my mind, one could apply bold to a few words, then italic to other words, and then use the style editor to un-bold the B, un-italic the I and just apply different colors.

But maybe it is wierd... How would you let the user do something like "a few words in red here, a few words in blue here" on 1 line?


> Also that mobile icon should be only indicator that will show on what mode you are, so when you are making a style for mobile you need to select mobile editor button to go into that mode, then everything you change from style will be on visible mobile only. I don't understand this concept you made now with mobile only?

Ok so the button should not be clickable?

In Silex, the styles you change in mobile mode, e.g. a background color, is not visible in desktop, but the styles you change in desktop are visible in mobile. Feel free to try to play with background colors in mobile and desktop modes

> Yes we need, that's why I made that accordion button "Layout" above Typography :) Just look at Webflow Layout section, that is just what we need.

Ok, same as before here: be patient, we'll get there :)

> I must admit I don't understand how this styling working. That's why I thought it should be the simple as possible and to use presented system we already can see how it works (Webflow for example) so it can be a model for Silex.
> I think it should automatically apply new style when add new text box, also when select text box on canvas it should show selected style for that box in style editor.

Yes ok, I tried that (it is also a bit like grapejs right?)
I have found it complicated to understand once I saw it working (it was the version you said you did not have time to test it)

I can do that again at least to test... I'll try to do that soon


---

Just for you to know, now that there is a start of a prototype and the UI design is integrated, I need to spend some time refactoring...

Thank you for your feedback guys! And for helping answering the other issues for "simple" questions, it give me more time to debug and work on this.

**Comment by lexoyo** [comment]
great videos @budislav thx
what do you think @ceubri ? this is very close to what i wanted to do at first

**Comment by ceubri** [comment]
Great !! And specially the vokoscreen-2018-05-04_18-00-03.mp4 and the layout part ;)
I try recently https://silex-test-tmp.herokuapp.com/ and it looks better than couple of weeks, good job Lexa, it's usefull.

**Comment by lexoyo** [comment]
Merci ceubri!!
So I finished my "refactoring" and even if I am not at all happy with it I have moved forward

Now on [my test instance](https://silex-test-tmp.herokuapp.com/) I have put something a little bit closer to your videos @budislav : when the user selects a text box, if it has a style applied to it, it will be selected in the style editor. The idea is to have a "all texts" style and then each text box can have a specific style

Try it like this:
* open https://silex-test-tmp.herokuapp.com/
* choose the "desktop" blank template (for some reason the mobile one doesn't work)
* open the style editor and add a style (this will be done by default in all the templates but it is not the case yet)
* call your style `Body initial`
* add a text box, play with this first style
* add a new style with any name and apply it to 1 or many new text boxes
* play with this new style and see how it overrides the 1st style
* select different text boxes and see how the style editor reacts

Then let me know how you feel about this UX?

Thank you for your help!!

**Comment by budislav** [comment]
@lexoyo This doesn't work good.
If I understand you well user need to make a style name for text first, then add a text box?
It should set a style name automatically when add a new text box every time. So one default text style and name for example called "text".
When I add a new box it should have default text style settings not the same as previous. But it will be automatically named text1.
Text box will have the same settings and same style name of text box only when you select than copy and paste it, or when you select a style from select style button.
Now if i want to change that particular text box to have a different style I will add a new style with add button, name it and it should automatically apply that new style settings I made to that style.
It is not working like that in your test. I need to press on apply button each first time which is not good.
It is good that it will make directly every change on text box on canvas in the moment I edit something, that is good!
But if I now want to select previous style I made and automatically apply to that button which is selected on canvas it should be easy. I select wanted style from select button and I see those settings are changed in the right panel which is awesome. But it wont apply that style on text box on canvas, because every time I select some style from select button it will deselect that text box on canvas so I can't even apply that style with button but it should be automatically. So it is not working good.
I hope you understand everything please ask me if you don't, this is really important to make it to work right.
Please to understand how it should work, look at these two short videos about webflow classes, I couldn't explain better than these videos.

https://www.youtube.com/watch?v=zmxo1ywQdVk
https://www.youtube.com/watch?v=LPABhi7uC6A

**Comment by lexoyo** [comment]
Ok so what we want is
* By default (body selected) => edit the style of all text boxes
* Select a text box without a style => edit a generated class name "Text Box 1"
* Select an element with a style => edit this style
* Click on the styles drop-down list and select another style => apply this style to the selected text box(es)

Is this correct? Did I miss something?

Remaining questions
* If I select multiple text boxes with different styles, what should happen?
* If I apply a style to a text box which has been styled, should it loose it's initial styles?
* If I select a text box and then select the "all text style", it would apply it? I should prevent it?

**Comment by lexoyo** [comment]
Style editor step 2:
![style-editor-step2](https://user-images.githubusercontent.com/715377/39960390-8a6b7d2a-5622-11e8-885c-3f8753b0badb.gif)

**Comment by budislav** [comment]
> Ok so what we want is
> 
>     By default (body selected) => edit the style of all text boxes
>     Select a text box without a style => edit a generated class name "Text Box 1"
>     Select an element with a style => edit this style
>     Click on the styles drop-down list and select another style => apply this style to the selected text box(es)
> 
> Is this correct? Did I miss something?
> 

I don't quite understand this points well so I will try to make scenario and how it should work:

1. User open blank template.
2. He add text element from add elements panel (Silex will auto name that class Text Box1 and every time when user insert new text box Silex will auto increment that number so it will be text box2, text box3,.. But he can rename it to whatever he want).
3. So he made one text box and he want to change his style. No problem whatever he change color line height,.. everything will auto change and show on canvas. Without pressing apply button, **we don't need that button**. Also everything he change will auto save on that style (class name) in this example text box1.
4. Now user add a new text box from add elements. (silex will auto name it text box2) but he want to change this name so he rename current style to Big title.  Settings he see for that new text box are not the same as previous textbox1, but they are default textbox settings, so everything he change will affect only that style he made.
5. User now want to change those style settings like color and other stuff. **While that text box is selected whatever** he change for example color that will **automatically show on canvas** and it will be **saved to that style Big title**. **Without touching apply button**. So previous textbox1 have it's own settings, nobody want to change that textbox1, user now changing Big title (textbox2)
6. User now copy and paste that big title text box couple of times and he have 3 big titles now. Every one oh them have the same style name "Big title" automatically.
7. Now everything user change with that big title will have same for all big title text boxes (3 of them). For example he touch any of that big title and change text color to red, everyone of them will have red color because they are under same style.
8. Now user want to change the look of one of big text to look of textbox1 (the first text box he made), so while that text box is selected on canvas one of those big title he can just select that style from style select button to textbox1 (because we did't renamed that class, that is auto name). And that big title will have same look like textbox1 and it will get that class textbox1. So that is belong to that class now (textbox1) not to big title class (style).
9. Now everything what he click on canvas any text box for example if he click on one of those 3 big titles now only 2 because we changed one, that style of selected box will appear in style editor, **automatically**.
10. Now user rename that textbox1 style to "small text" now those two text boxes he have with same style would be named small text.

> If I select multiple text boxes with different styles, what should happen?

This is a good question.
So if we want that to select multi text boxes with different styles we can change them but to only one style.
So if we select three text boxes for example and each of them have different styles (classes) or only two different styles in the moment when user select them, all settings in style editor should disappear and there should be a sentence in style editor "merge styles to one" and button under that "merge". So if user click on merge Silex would automatically made one style for all three text boxes named textboxmerged and incremented number if there are already merged text boxes before. This is just naming, so there should be different name always so not to have same settings as other styles.
So on the end user will have three boxes with one default text box style, simple as that. 
If user select on canvas two or more text boxes with **same styles** there are **no message** in style editor because they are having one same style. They share same style. So now those three text boxes share one style but if you user want he can just select a style from select style button and all text boxes will have that same style.

> If I apply a style to a text box which has been styled, should it loose it's initial styles?

Yes. They are sharing same style so if you change that text box that have same style as other boxes all boxes will have those new settings. But that's why Webflow have combo classes, like you can see in second video I sent you combo classes are ways to have a new style for that text box you want without worries you will destroy other text boxes. So it should be a way to add a new name to that style (class) you want to change like "combo" and it automatically made a new class. Please try to play with that in Webflow, you will get idea how it working. The most important is to have understandable clean html and css, so we need no complications.

> If I select a text box and then select the "all text style", it would apply it? I should prevent it?

I don not understand this option, I think we don't need that. Like I explained if user select one text box with style that all other text boxes have you basically selected everything because they are under the same style but you will see only those settings in style editor they all share not selected all text boxes on canvas. If we have for example 20 text boxes with all different styles and you want to change them all we can select on canvas all of them we want to change, we don't need that option because there should be only names of classes (styles) but all style is not a class. So there should not be hat option to select them all at all. Again please try to play with Webflow you will understand everything within couple of minutes I promise.

**Comment by budislav** [comment]
Yes! It working good well done!!

**Comment by budislav** [comment]
@lexoyo it should automatically name new added textbox, now it gives him a name but only after I made some changes. It should give that textbox name right after he add it on canvas.

**Comment by budislav** [comment]
@lexoyo yeah I think grouped styles don't working good. If you can please read my big answer above I think text boxes with different styles should merge into one, so no groups. That is much easier for user to understand and work.

**Comment by lexoyo** [comment]
Ok thx again @budislav for the quick answers
I'll do that soon (group styles)
I just published a version with mobile editor mode = mobile style (as you requested previously)

**Comment by budislav** [comment]
@lexoyo I just tried responsive template and there is a bug with styles you can see on this video it makes a text style6 but there is only one style.
[vokoscreen-2018-05-12_21-27-55.zip](https://github.com/silexlabs/Silex/files/1997875/vokoscreen-2018-05-12_21-27-55.zip)

**Comment by lexoyo** [comment]
Hi Budislav, thank you for the test! 

I did not manage to reproduce your bug but I had another bug when I clicked the "" style in the list. I fixed it and it may fix your bug too, could you test please? :-/

Also concerning the "style groups", this is just a default name created when you style several text boxes. I think you are talking about multiple classes in webflow? This I did not do yet

**Comment by budislav** [comment]
@lexoyo  I just tried on windows same here,
it makes 65 styles :/ 
Did you tried on other machine or only on your pc?
This is on Firefox I didn't tried other browsers.
![silexbug](https://user-images.githubusercontent.com/3619927/40008696-3827821e-57a0-11e8-8adc-728cd03dcf48.png)

**Comment by lexoyo** [comment]
> Did you tried on other machine or only on your pc?

I tried on firefox and chrome
Did someone else have such a bug? @ceubri @BaguetteAgile @rapenne-s ?

**Comment by lexoyo** [comment]
@budislav I don't know how to go forward with this
Maybe you could open the browser's developer console and tell me if there are any warnings or errors?

**Comment by budislav** [comment]
@lexoyo yes, this is all I see there
![untitled](https://user-images.githubusercontent.com/3619927/40247134-af33916c-5acb-11e8-9bd9-036497bc1cfd.png)
I really believe others can produce this because I tried it on two fresh installed OS but always on Firefox 60.0.1 (64bit)
I just tried in chromium and can't reproduce but in firefox always.
Do you have an access to some other pc, maybe to try in vm?

**Comment by budislav** [comment]
Also this
![untitled1](https://user-images.githubusercontent.com/3619927/40247504-f3e731e6-5acc-11e8-8cbf-7ce96ec4f9d2.png)

**Comment by lexoyo** [comment]
Did you apply styles with the developer tools open? I suppose that you did :(

I have 2 fedora computers and can not reproduce :'(

**Comment by lexoyo** [comment]
Ok I'll try again
Yes the address is correct

**Comment by lexoyo** [comment]
@budislav just an idea like that: did you empty your browser cache?

**Comment by lexoyo** [comment]
Yes please

**Comment by budislav** [comment]
Same problem :(
Now it made 91 styles for one box.

**Comment by lexoyo** [comment]
Anyone can help us test on other computers?

**Comment by BaguetteAgile** [comment]
Hello everyone, I'm back again.
I just tested the text editor on Linux with Web Chromium, no problem for me.

> So it appears that I would rather have the style editor NOT take the selection into account. Then we would have to create styles with a "+" button, and apply it with an "apply" button.

it seems to me that this solution would be a good option to be able to edit differently 2 blocks of the same name.

**Comment by budislav** [comment]
> or add * beside P?
I would say this, but this anyway should be changed in future if we are going to have components, and their settings.

> Here, what I often do is to change the font size, not realizing that it automatically creates a new style because a text box is selected... What I wanted to do is press esc key and then change the font size

If user want to select one particular text box and change it he can do that with making a new style I don't see this problem and if user need to select object and apply style every time that would be the hell for them and step back in development, but if you think it's better do what you need. Combo style from webflow is solution I think..
You don't need to erase that code completely if you can put in comment, that would be good.

**Comment by budislav** [comment]
I can't select a color anymore are you changed to old? I have more issue rigth now,but also my mouse is broken so he double click on something I don't want.

**Comment by lexoyo** [comment]
Releasing this now!
It will be live in 5 min, would someone have time to run some more tests on editor.silex.me ?
Next step the inline text editor
